palace/linalg/errorestimator.cpp

```diff
@@ -201,120 +296,177 @@ void CurlFluxErrorEstimator<VecType>::AddErrorIndicator(const VecType &U,
-
-      double elem_err = 0.0;
-      for (int i = 0; i < ir.GetNPoints(); i++)
+      CeedInt nsub_ops;
+      CeedOperator *sub_ops;
+      PalaceCeedCall(
+          ceed, CeedCompositeOperatorGetNumSub(integ_op[utils::GetThreadNum()], &nsub_ops));
+      PalaceCeedCall(
+          ceed, CeedCompositeOperatorGetSubList(integ_op[utils::GetThreadNum()], &sub_ops));
+      MFEM_ASSERT(nsub_ops > 0, "Unexpected empty libCEED composite operator!");
+      CeedOperatorField field;
+      PalaceCeedCall(ceed, CeedOperatorGetFieldByName(sub_ops[0], "curl_u_1", &field));
+      PalaceCeedCall(ceed, CeedOperatorFieldGetVector(field, &U_gf_vec));
+      PalaceCeedCall(ceed, CeedOperatorGetFieldByName(sub_ops[0], "u_2", &field));
+      PalaceCeedCall(ceed, CeedOperatorFieldGetVector(field, &F_gf_vec));
+      if constexpr (std::is_same<VecType, ComplexVector>::value)
       {
-        const mfem::IntegrationPoint &ip = ir.IntPoint(i);
-        T.SetIntPoint(&ip);
-        fe.CalcVShape(ip, Interp);
-        fe.CalcCurlShape(ip, Curl);
-        const double w = ip.weight * T.Weight();
-
-        auto AccumulateError = [&](const Vector &U_gf_, const Vector &F_gf_)
-        {
-          // μ⁻¹ ∇ × U -- μ⁻¹ |J|⁻¹ J ∇ × Uᵣ
-          U_gf_.GetSubVector(dofs, loc_gf);
-          if (dof_trans.GetDofTransformation())
-          {
-            dof_trans.InvTransformPrimal(loc_gf);
-          }
-          Curl.MultTranspose(loc_gf, V_ip);
-          T.Jacobian().Mult(V_ip, V_smooth);
-          mat_op.GetInvPermeability(T.Attribute).Mult(V_smooth, V_ip);
-          V_ip *= 1.0 / T.Weight();
-
-          // Smooth flux -- J⁻ᵀ Fᵣ
-          F_gf_.GetSubVector(dofs, loc_gf);
-          if (dof_trans.GetDofTransformation())
-          {
-            dof_trans.InvTransformPrimal(loc_gf);
-          }
-          Interp.MultTranspose(loc_gf, V_tmp);
-          T.InverseJacobian().MultTranspose(V_tmp, V_smooth);
-
-          V_smooth -= V_ip;
-          elem_err += w * (V_smooth * V_smooth);
-          return w * (V_ip * V_ip);
-        };
-        if constexpr (std::is_same<VecType, ComplexVector>::value)
-        {
-          loc_norm2 += AccumulateError(U_gf.Real(), F_gf.Real());
-          loc_norm2 += AccumulateError(U_gf.Imag(), F_gf.Imag());
-        }
-        else
-        {
-          loc_norm2 += AccumulateError(U_gf, F_gf);
-        }
+        ceed::InitCeedVector(U_gf.Real(), ceed, &U_gf_vec, false);
+        ceed::InitCeedVector(F_gf.Real(), ceed, &F_gf_vec, false);
       }
-      h_estimates[e] = std::sqrt(elem_err);
+      else
+      {
+        ceed::InitCeedVector(U_gf, ceed, &U_gf_vec, false);
+        ceed::InitCeedVector(F_gf, ceed, &F_gf_vec, false);
+      }
+    }
+
+    // Each thread writes to non-overlapping entries of the estimates vector.
+    CeedVector estimates_vec;
+    ceed::InitCeedVector(estimates, ceed, &estimates_vec);
+
+    // Do the integration (both input vectors are passive). For the complex case, add sum of
```

Comment:
Passive here means that neither is modified by the operations correct? so there's no possibility of race conditions/can be easily vectorized.

Reply:
Exactly, inputs are always read-only anyway. We have an operator `d = f(u, v)` which isn't nicely representable with a single active input vector. So both are just labeled as "passive" inputs.